### ISSUE TITLE
Facing "java.lang.IllegalStateException: Could not find a valid Docker environment. Please see logs and check configuration" when pushing to DockerHub

### ISSUE THREAD

**Issue by pvtuan10** [testcontainers/testcontainers-java#902]
Hi, my project is using `testcontainers` for integration tests and the whole project is built and pushed as a Docker image to DockerHub.

I am facing the issue of `Caused by: java.lang.IllegalStateException: Could not find a valid Docker environment. Please see logs and check configuration` when building my image in Docker. Do you have any suggestion to solve this problem?

My Dockerfile: https://github.com/jenkinsci/remoting-kafka-plugin/blob/master/Dockerfile
Build failed log: https://hub.docker.com/r/jenkins/remoting-kafka-agent/builds/by4exxjryxm4jznnbtgpsxf/

**Comment by kiview** [comment]
You haven't setup a Docker environment in your Dockerfile. I'm not sure how Docker-in-Docker works in the DockerHub (or if it even works?), but you could start by installing Docker in your build image.

**Comment by pvtuan10** [comment]
Hi @kiview, I tried to install Docker in my build image but cannot run Docker in Docker although I followed the instruction here https://www.testcontainers.org/usage/inside_docker.html

I got this error `docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.`

I also tried to run `service docker status` after `service docker start` but got this message `Docker is not running`.

You can check my `Dockerfile` here https://github.com/pvtuan10/remoting-kafka-plugin/blob/bugfix/fix-agent-docker-build/Dockerfile

Do you have any suggestion?

**Comment by kiview** [comment]
Your Dockerfile is a multi stage build, so the second stage would not have
the daemon or the socket mounted.

However, the root problem seems to be, that Docker Hub autobuild does not
support Docker in Docker configurations.

You can try to find out if it's actually supported, else I fear, would need
to use a different CI service for the testing step.

Pham Vu Tuan <notifications@github.com> schrieb am Mo., 8. Okt. 2018, 17:56:

**Comment by pvtuan10** [comment]
Yes, I also guess that Docker Hub does not support Docker in Docker configurations. The build cannot be successful even when I tried to run `hello-world` container with mounting volume `RUN docker run --rm -v $PWD:$PWD -w $PWD -v /var/run/docker.sock:/var/run/docker.sock hello-world`.

Thank you for your suggestion, and I will find another way to run my CI.

**Comment by kiview** [comment]
Yes, therefore I recommend TravisCI or CircleCI, which both support Docker out of the box and are free for Open Source usage, for building, as well as testing. You can still push the resulting image to the DockerHub.

**Comment by muzir** [comment]
I had some problem in CircleCI.

Use

```
build-test:
    machine: true
```
instead of
```
  build-test:
    docker:
      # specify the version you desire here
      - image: circleci/openjdk:8-jdk
```

Check below link to get more information about machine usage in CircleCI.

https://circleci.com/docs/2.0/executor-types/#using-machine
